File: python-backend/backend/types/events.py
"""Pydantic models for viral events, deep dives, and audits."""

from pydantic import BaseModel, Field

from backend.types.enums import ConfidenceLevel, EventCategory, WhyIncludedTag


class ViralEvent(BaseModel):
    headline: str
    category: EventCategory
    confidence: ConfidenceLevel
    rank: int = Field(default=0)
    country_region: str = Field(default="")
    why_included: list[WhyIncludedTag] = Field(default_factory=list)
    revenue_impact: str = Field(default="")
    what_changed: list[str] = Field(default_factory=list)
    proof_pack: str = Field(default="")
    # Backward compat: old fields kept as optional
    impact_rating: int | None = Field(default=None)
    source: str | None = Field(default=None)
    summary: str | None = Field(default=None)


class DeepDive(BaseModel):
    title: str
    what_happened: str = Field(default="")
    why_it_matters: str = Field(default="")
    second_order_implications: str = Field(default="")
    what_to_watch: str = Field(default="")
    # Backward compat: old fields kept as optional
    priority: str | None = Field(default=None)
    summary: str | None = Field(default=None)
    key_findings: list[str] | None = Field(default=None)


class CompletenessAudit(BaseModel):
    verified_signals: int
    sources_checked: int
    confidence_score: float = Field(ge=0.0, le=1.0)
    gaps: list[str]
    reuters_articles: list[str] = Field(default_factory=list)
    major_stock_moves: list[str] = Field(default_factory=list)
    vendor_coverage: list[str] = Field(default_factory=list)

File: rust-deepagent-backend/src/types/events.rs
use serde::{Deserialize, Serialize};

use super::enums::{ConfidenceLevel, EventCategory, WhyIncludedTag};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViralEvent {
    pub headline: String,
    pub category: EventCategory,
    pub confidence: ConfidenceLevel,
    #[serde(default)]
    pub rank: i32,
    #[serde(default)]
    pub country_region: String,
    #[serde(default)]
    pub why_included: Vec<WhyIncludedTag>,
    #[serde(default)]
    pub revenue_impact: String,
    #[serde(default)]
    pub what_changed: Vec<String>,
    #[serde(default)]
    pub proof_pack: String,
    // Backward compat: old fields kept as Option so legacy reports still deserialize
    #[serde(default)]
    pub impact_rating: Option<i32>,
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeepDive {
    pub title: String,
    #[serde(default)]
    pub what_happened: String,
    #[serde(default)]
    pub why_it_matters: String,
    #[serde(default)]
    pub second_order_implications: String,
    #[serde(default)]
    pub what_to_watch: String,
    // Backward compat: old fields kept as Option
    #[serde(default)]
    pub priority: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub key_findings: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompletenessAudit {
    pub verified_signals: i32,
    pub sources_checked: i32,
    pub confidence_score: f64,
    pub gaps: Vec<String>,
    #[serde(default)]
    pub reuters_articles: Vec<String>,
    #[serde(default)]
    pub major_stock_moves: Vec<String>,
    #[serde(default)]
    pub vendor_coverage: Vec<String>,
}
